### PR TITLE
Confirm home replacement when overwriting existing homes

### LINKED ISSUE #2038: SUGGESTION prompt before setting home with the same name.
Would be super cool if we had a prompt that asks you to confirm when you're setting a home name that you've already set. (confirmation before overwriting)

thank you

### PR DESCRIPTION
**Pull request by pop4959**
Warns the player that they are trying to replace a home, if they try to set one with the same name, when `confirm-home-overwrite` is set to `true` in the config.

image

Fixes #2038
Closes #2847

### CODE REVIEW

**Review by mdcfe — APPROVED** [review]

**Comment by pop4959** [comment]
Minor nitpick, but do we want the warning to expire at some point? Currently, a player attempts to overwrite a home name, they will be warned, but if they don't set any other home, they will not be warned again about it unless the server restarts. Not sure if it matters all too much since hopefully one warning is enough for most people to realize how /sethome works, but I could also add in a timestamp and let the warning appear again if > 1 minute has passed.

**Review by mdcfe — CHANGES_REQUESTED** [review]
I think we should track which home the confirmation was shown for, so that if they try to overwrite a different home it re-prompts the user. I also agree a timeout (1/2 minutes?) would be a good idea.

**Comment by pop4959** [comment]
Confirmation now appears if either of the two conditions are met:
1) The user is trying to overwrite a home, and it's different from the last that they tried to set.
2) The user is trying to overwrite a home, and it's the same one, but it has been too long (2+ minutes) since the last confirmation.

**Comment by ToonTown0909** [comment]
Is this something that can be set as a toggle, as opposed to permanently? (ex; /tptoggle, /msgtoggle, /paytoggle)

**Comment by pop4959** [comment]
> Is this something that can be set as a toggle, as opposed to permanently? (ex; /tptoggle, /msgtoggle, /paytoggle)

I can see the reasoning, but I don't think there is much of a point, because this is not something that happens often, and when it does, it's more often than not a mistake. I added a configuration value just so that server owners can choose not to use the new warnings.

Edit: Maybe this could also be a permission instead. I'm assuming you're asking because you would want to disable this for yourself / admins who know better, but keep it enabled for your players.

File: Essentials/src/com/earth2me/essentials/commands/Commandsethome.java [hunk 1]
```diff
@@ -55,6 +56,13 @@ public void run(final Server server, final User user, final String commandLabel,
...
 
+        if (ess.getSettings().isConfirmHomeOverwrite() && usersHome.hasHome(name) && (!name.equals(usersHome.getLastHomeConfirmation()) || name.equals(usersHome.getLastHomeConfirmation()) && System.currentTimeMillis() - usersHome.getLastHomeConfirmationTimestamp() > TimeUnit.MINUTES.toMillis(2))) {
+            usersHome.setLastHomeConfirmation(name);
+            usersHome.setLastHomeConfirmationTimestamp();
+            user.sendMessage(tl("homeConfirmation", name));
+            return;
+        }
+
         usersHome.setHome(name, location);
         user.sendMessage(tl("homeSet", user.getLocation().getWorld().getName(), user.getLocation().getBlockX(), user.getLocation().getBlockY(), user.getLocation().getBlockZ(), name));
```
Comment (JRoy) [inline Essentials/src/com/earth2me/essentials/commands/Commandsethome.java:67]:
```suggestion
        user.sendMessage(tl("homeSet", user.getLocation().getWorld().getName(), user.getLocation().getBlockX(), user.getLocation().getBlockY(), user.getLocation().getBlockZ(), name));
        usersHome.setLastHomeConfirmation(null);
```

File: Essentials/src/messages.properties [hunk 2]
```diff
@@ -182,6 +182,7 @@ holdFirework=\u00a74You must be holding a firework to add effects.
 holdPotion=\u00a74You must be holding a potion to apply effects to it.
 holeInFloor=\u00a74Hole in floor\!
 homes=\u00a76Homes\:\u00a7r {0}
+homeConfirmation=\u00a76You already have a home named \u00a7c{0}\u00a76!\nType the command again to confirm replacement.
```
Comment (JRoy) [inline Essentials/src/messages.properties]: Remain consistent style with other confirm messages

```suggestion
homeConfirmation=\u00a77To \u00a7lCONFIRM\u00a77 overwriting your existing home, \u00a7c{0}\u00a77, please repeat the command.
```

**Review by JRoy — APPROVED** [review]

**Review by JRoy — DISMISSED** [review]
Didn't mean to approve, check my review comments

File: Essentials/src/messages.properties (hunk 2 above)
Comment (pop4959) [reply Essentials/src/messages.properties]: If anything, the style and format of that message varies wildly from the rest of Essentials. I'm not saying the message in this PR is perfect, but I think that one is also a bit far off the mark.

**Comment by mibby** [comment]
Feature seems to work great. Does what it accomplishes without any issues. Timeout also functions, although no message to let players know it timed out.

**Comment by pop4959** [comment]
@mibby It's like that on purpose, because it would be rather strange as a player to receive a notification about the warning being reset. The player only cares about the warning in the moment.

**Review by mdcfe — DISMISSED** [review]

File: Essentials/src/messages.properties (hunk 2 above)
Comment (mdcfe) [reply Essentials/src/messages.properties]:
```suggestion
homeConfirmation=\u00a76You already have a home named \u00a7c{0}\u00a76!\nTo overwrite your existing home, type the command again.
```

**Review by mdcfe — APPROVED** [review]

**Comment by pop4959** [comment]
Sneaky, but I like your message better so it's fine

**Pull request merged**